sn_client/src/audit/dag_crawling.rs

```diff
@@ -62,26 +64,93 @@ impl Client {
         spend_processing: Option<Sender<SignedSpend>>,
         verify: bool,
     ) -> WalletResult<SpendDag> {
-        info!("Building spend DAG from {spend_addr:?}");
-        let mut dag = SpendDag::new(spend_addr);
+        let (tx, mut rx) = tokio::sync::mpsc::channel(SPENDS_PROCESSING_BUFFER_SIZE);
+
+        // start crawling from the given spend address
+        let self_clone = self.clone();
+        let crawl_handle =
+            tokio::spawn(async move { self_clone.spend_dag_crawl_from(spend_addr, tx).await });
+
+        // start DAG building from the spends gathered while crawling
+        // forward spends to processing if provided
+        let build_handle: tokio::task::JoinHandle<Result<SpendDag, WalletError>> =
+            tokio::spawn(async move {
+                let mut dag = SpendDag::new(spend_addr);
+                while let Some(spend) = rx.recv().await {
+                    let addr = spend.address();
+                    dag.insert(addr, spend.clone());
+                    if let Some(sender) = &spend_processing {
+                        sender
+                            .send(spend)
+                            .await
+                            .map_err(|e| WalletError::SpendProcessing(e.to_string()))?;
+                    }
+                }
+                Ok(dag)
+            });
+
+        // wait for both to finish
+        let (crawl_res, build_res) = tokio::join!(crawl_handle, build_handle);
+        crawl_res.map_err(|e| {
+            WalletError::SpendProcessing(format!("Failed to Join crawling results {e}"))
+        })??;
+        let mut dag = build_res.map_err(|e| {
+            WalletError::SpendProcessing(format!("Failed to Join DAG building results {e}"))
+        })??;
+
+        // verify the DAG
+        if verify {
+            info!("Now verifying SpendDAG from {spend_addr:?} and recording errors...");
+            let start = std::time::Instant::now();
+            if let Err(e) = dag.record_faults(&dag.source()) {
+                let s = format!(
+                    "Collected DAG starting at {spend_addr:?} is invalid, this is probably a bug: {e}"
+                );
+                error!("{s}");
+                return Err(WalletError::Dag(s));
+            }
+            let elapsed = start.elapsed();
+            info!("Finished verifying SpendDAG from {spend_addr:?} in {elapsed:?}");
+        }
+
+        Ok(dag)
+    }
+
+    /// Crawls the Spend Dag from a given SpendAddress recursively
+    /// following descendants all the way to UTXOs
+    /// Returns the UTXOs reached
+    pub async fn spend_dag_crawl_from(
```

Comment:
shall be feature guarded as well?

Reply:
This is `sn_client` code, the feature guard is a `sn_auditor` option, the lib shall have no feature guard.